### PR TITLE
Remove bv support

### PR DESCRIPTION
**Pull request by makaimann**
This PR removes support for BV which simplifies the code a lot, and makes a few other clean up changes. For example, removing support for multiple read/write UF and adding an `abstract` and `concrete` function which can translate between the concrete system and the abstraction.

Unfortunately, it looks like we lose a few benchmarks overall on this branch (but also gain some unique solves). It still seems pretty volatile just based on luck I think with predicates and refinement axioms. I was thinking we could deal with that later, because as we add other refinement options, this codebase will change a lot more anyway. What do you think @ahmed-irfan?

Screenshot from 2020-06-08 17-15-42

### CODE REVIEW

File: prophic3/array_axiom_enumerator.h [hunk 1]
```diff
@@ -16,50 +16,33 @@ class ArrayAxiomEnumerator {
...
 
       if (!ts.contains_next(idx))
       {
-        curr_indices_[typestr].insert(idx);
-        curr_indices_no_witnesses_[typestr].insert(idx);
+        curr_indices_.insert(idx);
+        curr_indices_no_witnesses_.insert(idx);
       }
-      orig_indices_[typestr].insert(idx);
+      orig_indices_.insert(idx);
```
Comment (ahmed-irfan) [inline prophic3/array_axiom_enumerator.h:37]: what is the difference between orig_indices_ and orig_indices_set_?

File: prophic3/prophic3.cpp [hunk 2]
```diff
@@ -1014,9 +993,26 @@ void ProphIC3::print_witness(msat_model model,
...
   logger(1) << "+++++++++++++++++++++ UF values ++++++++++++++++++++"
             << endlog;
-  for (auto elem : abstractor.read_ufs()) {
-    msat_term arr = elem.first;
-    msat_decl fun = elem.second;
+
+  // collect all the arrays
+  TermList abs_arrays;
+  for (auto sv : abs_ts_.statevars()) {
+    msat_term concsv = abstractor.concrete(sv);
```
Comment (ahmed-irfan) [inline prophic3/prophic3.cpp:1000]: can you use get_orig_type here?

File: prophic3/prophic3.cpp [hunk 3]
```diff
@@ -1014,9 +993,26 @@ void ProphIC3::print_witness(msat_model model,
...
+    msat_type typ = msat_term_get_type(concsv);
+    if (msat_is_array_type(msat_env_, typ, nullptr, nullptr)) {
+      abs_arrays.push_back(sv);
+    }
+  }
+  for (auto iv : abs_ts_.inputvars()) {
+    msat_term conciv = abstractor.concrete(iv);
```
Comment (ahmed-irfan) [inline prophic3/prophic3.cpp:1007]: ditto

**Comment by ahmed-irfan** [comment]
we are getting more errors. what are the errors?

File: prophic3/array_axiom_enumerator.h (hunk 1 above)
Comment (makaimann) [reply prophic3/array_axiom_enumerator.h:37]: Good catch! There's no difference anymore. Changed with: 2b4ad50fb7796987b1570987ab179c988496b3cd

By the way, I think I'll also get rid of most of these different index sets and instead detect the indices on the fly like you've mentioned before. I'll make that in a separate PR.

File: prophic3/prophic3.cpp (hunk 2 above)
Comment (makaimann) [reply prophic3/prophic3.cpp]: Yep! Good catch. Changed with: 43a7b4464767d6baf8d5cbe4cfb8b307645ba56c

File: prophic3/prophic3.cpp (hunk 3 above)
Comment (makaimann) [reply prophic3/prophic3.cpp]: Good catch! Changed with: 43a7b4464767d6baf8d5cbe4cfb8b307645ba56c

**Comment by makaimann** [comment]
I checked and 9 of the errors are "IC3 returned undefined", e.g. it failed to find new predicates in refinement. And the other 11 are all essentially timeouts. I checked this by seeing if the error file has any lines that don't start with `runlim`. The only ones that don't are the 9 exceptions mentioned earlier. I think it's just in the middle of IC3 and doesn't respond to the interrupt. There's probably a way to fix this, I think I can set an interrupt handler or something

File: prophic3/array_axiom_enumerator.h (hunk 1 above)
Comment (ahmed-irfan) [reply prophic3/array_axiom_enumerator.h:37]: cool.

**Pull request merged**